File: import_tasks_from_csv.py
"""
CSV fayldan task'larni va stage assignment'larni import qilish
Bu skript:
1. CSV'dan task ma'lumotlarini o'qiydi
2. Client, Branch, User'larni topadi yoki yaratadi
3. Task'larni topadi yoki yaratadi
4. Stage assignment'larni update qiladi
"""
import csv
import sys
from pathlib import Path
from datetime import datetime
import paramiko

# CSV fayl yo'li
csv_path = Path(r"c:\Users\11870\OneDrive\Desktop\Chiqim_turlari.csv")

# Stage nomlari mapping (CSV dagi nomlar -> Database dagi nomlar)
STAGE_MAPPING = {
    'Invoys': 'Invoys',
    'Zayavka': 'Zayavka',
    'TIR-SMR': 'TIR-SMR',
    'ST': 'ST',
    'FITO': 'Fito',
    'Deklaratsiya': 'Deklaratsiya',
    'Xujjat_tekshirish': 'Tekshirish',
    'Xujjat_topshirish': 'Topshirish',
    'Pochta': 'Pochta',
}

# Stage order mapping
STAGE_ORDER = {
    'Invoys': 1,
    'Zayavka': 2,
    'TIR-SMR': 3,
    'ST': 4,
    'Fito': 5,
    'Deklaratsiya': 6,
    'Tekshirish': 7,
    'Topshirish': 8,
    'Pochta': 9,
}

def parse_csv():
    """CSV faylni o'qib, task va stage ma'lumotlarini qaytaradi"""
    tasks = []
    
    # Encoding'ni tekshirish
    encodings = ['utf-8-sig', 'windows-1251', 'cp1251', 'utf-8', 'latin-1']
    encoding_used = None
    
    for enc in encodings:
        try:
            with open(csv_path, 'r', encoding=enc) as f:
                f.read(100)  # Test read
            encoding_used = enc
            break
        except (UnicodeDecodeError, UnicodeError):
            continue
    
    if not encoding_used:
        raise ValueError("CSV fayl encoding'ini aniqlab bo'lmadi")
    
    print(f"Fayl encoding: {encoding_used}")
    
    with open(csv_path, 'r', encoding=encoding_used) as f:
        reader = csv.DictReader(f, delimiter=';')
        
        for row in reader:
            task_id = row.get('Task ID', '').strip()
            if not task_id:
                continue
            
            # Task asosiy ma'lumotlari
            task_data = {
                'task_id': task_id,
                'client_name': row.get('Klient', '').strip(),
                'task_name': row.get('Task Name', '').strip(),
                'branch_name': row.get('Filial', '').strip(),
                'psr': row.get('PSR', '').strip(),
                'status': row.get('Status', '').strip(),
                'driver_phone': row.get('Shopir Tel raqami', '').strip(),
                'deal_amount': row.get('Kelishuv summasi', '').strip(),
                'expenses': row.get('Rasxodlar', '').strip(),
                'creation_date': row.get('Creation Date', '').strip(),
            }
            
            # Stage assignment'lar
            stage_assignments = []
            
            for csv_stage_name, db_stage_name in STAGE_MAPPING.items():
                stage_status = row.get(csv_stage_name, '').strip()
                worker_name = row.get(f'{csv_stage_name}_bajaruvchi', '').strip()
                
                if stage_status == 'Tayyor' and worker_name:
                    stage_assignments.append({
                        'stage_name': db_stage_name,
                        'worker_name': worker_name,
                        'status': 'TAYYOR',
                    })
            
            task_data['stage_assignments'] = stage_assignments
            tasks.append(task_data)
    
    return tasks

def generate_sql_script(tasks):
    """SQL script yaratadi"""
    sql_lines = []
    sql_lines.append("-- CSV fayldan task va stage assignment import qilish")
    sql_lines.append("-- Bu scriptni database'da bajarishdan oldin backup oling!")
    sql_lines.append("")
    sql_lines.append("BEGIN;")
    sql_lines.append("")
    
    # User name mapping (CSV dagi ismlar -> Database dagi user ID'larni topish uchun)
    sql_lines.append("-- Avval user'larni name bo'yicha topamiz")
    sql_lines.append("DO $$")
    sql_lines.append("DECLARE")
    sql_lines.append("    v_user_id INTEGER;")
    sql_lines.append("    v_task_id INTEGER;")
    sql_lines.append("    v_stage_id INTEGER;")
    sql_lines.append("    v_client_id INTEGER;")
    sql_lines.append("    v_branch_id INTEGER;")
    sql_lines.append("BEGIN")
    sql_lines.append("")
    
    processed_tasks = 0
    
    for task in tasks:
        # Client va Branch'ni topish/yaratish
        sql_lines.append(f"    -- Task: {task['task_name'][:50]} (ID: {task['task_id']})")
        sql_lines.append(f"    -- Client: {task['client_name']}, Branch: {task['branch_name']}")
        sql_lines.append("")
        
        # Branch'ni topish
        sql_lines.append(f"    SELECT id INTO v_branch_id FROM \"Branch\" WHERE name = '{task['branch_name']}';")
        sql_lines.append("    IF v_branch_id IS NULL THEN")
        sql_lines.append(f"        INSERT INTO \"Branch\" (name) VALUES ('{task['branch_name']}') RETURNING id INTO v_branch_id;")
        sql_lines.append("    END IF;")
        sql_lines.append("")
        
        # Client'ni topish/yaratish
        sql_lines.append(f"    SELECT id INTO v_client_id FROM \"Client\" WHERE name = '{task['client_name']}';")
        sql_lines.append("    IF v_client_id IS NULL THEN")
        sql_lines.append(f"        INSERT INTO \"Client\" (name, phone, \"dealAmount\") VALUES ('{task['client_name']}', '{task['driver_phone']}', {task['deal_amount'] or 'NULL'}) RETURNING id INTO v_client_id;")
        sql_lines.append("    END IF;")
        sql_lines.append("")
        
        # Task'ni topish (title va clientId bo'yicha)
        # Agar topilmasa, yaratish kerak (lekin bu murakkab, shuning uchun faqat topilgan task'larni update qilamiz)
        task_name_escaped = task['task_name'].replace("'", "''")
        sql_lines.append(f"    SELECT id INTO v_task_id FROM \"Task\" WHERE title = '{task_name_escaped}' AND \"clientId\" = v_client_id LIMIT 1;")
        sql_lines.append("")
        
        # Agar task topilsa, stage assignment'larni update qilish
        sql_lines.append("    IF v_task_id IS NOT NULL THEN")
        
        for stage in task['stage_assignments']:
            worker_name = stage['worker_name'].replace("'", "''")
            stage_name = stage['stage_name'].replace("'", "''")
            
            sql_lines.append(f"        -- Stage: {stage_name}, Worker: {worker_name}")
            sql_lines.append(f"        SELECT id INTO v_user_id FROM \"User\" WHERE name = '{worker_name}' LIMIT 1;")
            sql_lines.append("        IF v_user_id IS NOT NULL THEN")
            sql_lines.append(f"            SELECT id INTO v_stage_id FROM \"TaskStage\" WHERE \"taskId\" = v_task_id AND name = '{stage_name}';")
            sql_lines.append("            IF v_stage_id IS NOT NULL THEN")
            sql_lines.append(f"                UPDATE \"TaskStage\" SET status = 'TAYYOR', \"assignedToId\" = v_user_id, \"completedAt\" = NOW() WHERE id = v_stage_id;")
            sql_lines.append("            END IF;")
            sql_lines.append("        END IF;")
            sql_lines.append("")
        
        sql_lines.append("    END IF;")
        sql_lines.append("")
        
        processed_tasks += 1
        if processed_tasks % 10 == 0:
            sql_lines.append(f"    -- Processed {processed_tasks} tasks")
            sql_lines.append("")
    
    sql_lines.append("END $$;")
    sql_lines.append("")
    sql_lines.append("COMMIT;")
    
    return "\n".join(sql_lines)

def main():
    try:
        print("CSV faylni o'qish...")
        tasks = parse_csv()
        print(f"Jami {len(tasks)} ta task topildi")
        
        print("\nSQL script yaratilmoqda...")
        sql_script = generate_sql_script(tasks)
        
        output_file = Path("import_tasks.sql")
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(sql_script)
        
        print(f"\nSQL script yaratildi: {output_file}")
        print(f"Jami {len(tasks)} ta task uchun SQL script tayyor")
        print("\nEslatma: Bu scriptni database'da bajarishdan oldin backup oling!")
        
    except Exception as e:
        print(f"Xatolik: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()

if __name__ == '__main__':
    main()

File: backend/scripts/import-tasks-from-csv.ts
/**
 * CSV fayldan task'larni va stage assignment'larni import qilish
 * 
 * Foydalanish:
 *   cd backend
 *   npx ts-node --transpile-only scripts/import-tasks-from-csv.ts
 */

import { PrismaClient } from '@prisma/client';
import * as fs from 'fs';
import * as path from 'path';

const prisma = new PrismaClient();

// Stage nomlari mapping (CSV dagi nomlar -> Database dagi nomlar)
const STAGE_MAPPING: Record<string, string> = {
  'Invoys': 'Invoys',
  'Zayavka': 'Zayavka',
  'TIR-SMR': 'TIR-SMR',
  'ST': 'ST',
  'FITO': 'Fito',
  'Deklaratsiya': 'Deklaratsiya',
  'Xujjat_tekshirish': 'Tekshirish',
  'Xujjat_topshirish': 'Topshirish',
  'Pochta': 'Pochta',
};

// Stage order mapping
const STAGE_ORDER: Record<string, number> = {
  'Invoys': 1,
  'Zayavka': 2,
  'TIR-SMR': 3,
  'ST': 4,
  'Fito': 5,
  'Deklaratsiya': 6,
  'Tekshirish': 7,
  'Topshirish': 8,
  'Pochta': 9,
};

interface TaskData {
  task_id: string;
  client_name: string;
  task_name: string;
  branch_name: string;
  psr: string;
  status: string;
  driver_phone: string;
  deal_amount: string;
  expenses: string;
  creation_date: string;
  stage_assignments: Array<{
    stage_name: string;
    worker_name: string;
    status: string;
  }>;
}

function parseDate(dateStr: string): Date | null {
  if (!dateStr) return null;
  // Format: "24.11.2025 19:57:29"
  const parts = dateStr.split(' ');
  if (parts.length < 1) return null;
  
  const dateParts = parts[0].split('.');
  if (dateParts.length !== 3) return null;
  
  const day = parseInt(dateParts[0], 10);
  const month = parseInt(dateParts[1], 10) - 1; // Month is 0-indexed
  const year = parseInt(dateParts[2], 10);
  
  let hours = 0, minutes = 0, seconds = 0;
  if (parts.length > 1) {
    const timeParts = parts[1].split(':');
    if (timeParts.length >= 1) hours = parseInt(timeParts[0], 10);
    if (timeParts.length >= 2) minutes = parseInt(timeParts[1], 10);
    if (timeParts.length >= 3) seconds = parseInt(timeParts[2], 10);
  }
  
  return new Date(year, month, day, hours, minutes, seconds);
}

async function importTasks() {
  // CSV fayl yo'li - Windows path
  const csvPath = 'c:/Users/11870/OneDrive/Desktop/Chiqim_turlari.csv';
  
  if (!fs.existsSync(csvPath)) {
    console.error(`CSV fayl topilmadi: ${csvPath}`);
    process.exit(1);
  }
  
  console.log('CSV faylni o\'qish...');
  
  // Encoding'ni tekshirish va o'qish
  let content: string;
  const encodings = ['utf-8-sig', 'windows-1251', 'cp1251', 'utf-8', 'latin-1'];
  
  for (const enc of encodings) {
    try {
      content = fs.readFileSync(csvPath, { encoding: enc as BufferEncoding });
      console.log(`Fayl encoding: ${enc}`);
      break;
    } catch (e) {
      continue;
    }
  }
  
  if (!content!) {
    throw new Error("CSV fayl encoding'ini aniqlab bo'lmadi");
  }
  
  // CSV'ni manual parse qilish
  const lines = content.split('\n').filter(line => line.trim());
  const headers = lines[0].split(';').map(h => h.trim());
  
  const records: Record<string, string>[] = [];
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(';');
    const record: Record<string, string> = {};
    headers.forEach((header, idx) => {
      record[header] = values[idx]?.trim() || '';
    });
    records.push(record);
  }
  
  console.log(`Jami ${records.length} ta qator topildi`);
  
  let processed = 0;
  let updated = 0;
  let notFound = 0;
  let errors = 0;
  
  // User name cache
  const userCache = new Map<string, number>();
  
  // Avval barcha user'larni cache'ga yuklash
  const users = await prisma.user.findMany({
    select: { id: true, name: true },
  });
  
  for (const user of users) {
    userCache.set(user.name, user.id);
  }
  
  console.log(`User cache: ${userCache.size} ta user`);
  
  for (const row of records) {
    const taskId = row['Task ID']?.trim();
    if (!taskId) continue;
    
    try {
      const taskData: TaskData = {
        task_id: taskId,
        client_name: row['Klient']?.trim() || '',
        task_name: row['Task Name']?.trim() || '',
        branch_name: row['Filial']?.trim() || '',
        psr: row['PSR']?.trim() || '',
        status: row['Status']?.trim() || '',
        driver_phone: row['Shopir Tel raqami']?.trim() || '',
        deal_amount: row['Kelishuv summasi']?.trim() || '',
        expenses: row['Rasxodlar']?.trim() || '',
        creation_date: row['Creation Date']?.trim() || '',
        stage_assignments: [],
      };
      
      // Stage assignment'larni to'plash
      for (const [csvStageName, dbStageName] of Object.entries(STAGE_MAPPING)) {
        const stageStatus = row[csvStageName]?.trim();
        const workerName = row[`${csvStageName}_bajaruvchi`]?.trim();
        
        if (stageStatus === 'Tayyor' && workerName) {
          taskData.stage_assignments.push({
            stage_name: dbStageName,
            worker_name: workerName,
            status: 'TAYYOR',
          });
        }
      }
      
      if (taskData.stage_assignments.length === 0) {
        continue; // Stage assignment'lar bo'lmasa, o'tkazib yuborish
      }
      
      // Branch'ni topish yoki yaratish
      let branch = await prisma.branch.findFirst({
        where: { name: taskData.branch_name },
      });
      
      if (!branch) {
        branch = await prisma.branch.create({
          data: { name: taskData.branch_name },
        });
        console.log(`  [NEW] Branch yaratildi: ${taskData.branch_name}`);
      }
      
      // Client'ni topish yoki yaratish
      let client = await prisma.client.findFirst({
        where: { name: taskData.client_name },
      });
      
      if (!client) {
        client = await prisma.client.create({
          data: {
            name: taskData.client_name,
            phone: taskData.driver_phone || null,
            dealAmount: taskData.deal_amount ? parseFloat(taskData.deal_amount) : null,
          },
        });
        console.log(`  [NEW] Client yaratildi: ${taskData.client_name}`);
      }
      
      // Task'ni topish (title va clientId bo'yicha)
      let task = await prisma.task.findFirst({
        where: {
          title: taskData.task_name,
          clientId: client.id,
        },
        include: {
          stages: true,
        },
      });
      
      if (!task) {
        // Task topilmasa, yaratish (admin user kerak)
        const adminUser = await prisma.user.findFirst({
          where: { role: 'ADMIN' },
        });
        
        if (!adminUser) {
          console.error(`  [ERROR] Admin user topilmadi, task yaratib bo'lmadi: ${taskData.task_name}`);
          errors++;
          continue;
        }
        
        const creationDate = parseDate(taskData.creation_date) || new Date();
        
        task = await prisma.task.create({
          data: {
            title: taskData.task_name,
            clientId: client.id,
            branchId: branch.id,
            createdById: adminUser.id,
            driverPhone: taskData.driver_phone || null,
            hasPsr: taskData.psr ? taskData.psr.toLowerCase().includes('ha') : false,
            snapshotDealAmount: taskData.deal_amount ? parseFloat(taskData.deal_amount) : null,
            createdAt: creationDate,
            stages: {
              create: Object.keys(STAGE_ORDER).map((stageName) => ({
                name: stageName,
                stageOrder: STAGE_ORDER[stageName],
                status: 'BOSHLANMAGAN',
              })),
            },
          },
          include: {
            stages: true,
          },
        });
        
        console.log(`  [NEW] Task yaratildi: ${taskData.task_name} (ID: ${task.id})`);
      }
      
      // Stage assignment'larni update qilish
      for (const stageAssignment of taskData.stage_assignments) {
        const workerId = userCache.get(stageAssignment.worker_name);
        
        if (!workerId) {
          console.warn(`  [WARN] Worker topilmadi: ${stageAssignment.worker_name}`);
          continue;
        }
        
        const stage = task.stages.find((s) => s.name === stageAssignment.stage_name);
        
        if (!stage) {
          console.warn(`  [WARN] Stage topilmadi: ${stageAssignment.stage_name} (Task ID: ${task.id})`);
          continue;
        }
        
        // Stage'ni update qilish
        await prisma.taskStage.update({
          where: { id: stage.id },
          data: {
            status: 'TAYYOR',
            assignedToId: workerId,
            completedAt: new Date(),
          },
        });
      }
      
      updated++;
      processed++;
      
      if (processed % 10 === 0) {
        console.log(`  Processed ${processed} tasks...`);
      }
      
    } catch (error: any) {
      console.error(`  [ERROR] Task ID ${taskId}: ${error.message}`);
      errors++;
    }
  }
  
  console.log('\n=== Import natijalari ===');
  console.log(`Jami qatorlar: ${records.length}`);
  console.log(`Ishlangan: ${processed}`);
  console.log(`Yangilangan: ${updated}`);
  console.log(`Xatolar: ${errors}`);
}

async function main() {
  try {
    await importTasks();
  } catch (error) {
    console.error('Xatolik:', error);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
  }
}

main();

